raster/r.viewshed/main.cpp

```diff
@@ -501,7 +501,7 @@ parse_args(int argc, char *argv[], int *vpRow, int *vpCol,
     obsElevOpt->required = NO;
     obsElevOpt->key_desc = "value";
     obsElevOpt->description = _("Viewing elevation above the ground");
-    obsElevOpt->answer = "1.75";
+    obsElevOpt->answer = strdup("1.75");
```

Comment:
String from strdup() should be freed. If this is a problem, an alternative solution must be found.

Reply:
The `strdup()` call is there likely to avoid the issue of assigning `const char *` to `char *` which in general can be problem and causes a "discarding const qualifiers" warning. I think that's the underlying problem here. The GRASS solution to that is `G_store` which I think suffers from the same issues as `strdup`.

Since this is setting a default which, if I'm not mistaken, is only read, not used for storage, it is safe to recast. In C++ (where this is a warning on Linux with GCC for a long time already), I was using `const_cast<char*>("...")` for the answer. I assume, but never needed to try, that in C, a `(char *)"..."` cast should do the trick, i.e., express that here it is indeed okay and intended to assign literal to `char *`.

I think the ultimate solution (to this and several other issues with these structures) would be a function-based API for the options rather than directly assigning to the member variables.

As a heads-up, the freeing will be/is an issue in many other places.

Reply:
I think explicitly free'ing the string from strdup() is not needed because the default answer is allocated only once and automatically free'd when the program terminates.

Reply:
> I believe this is not the place to start a discussion of memory handling in grass. That been said, It is my understanding that harmless memory leaks are accepted in GRASS (due to short program/module life) — that’s why I suggested it in the first place. I’d prefer strdup() or rather G_store (thanks, that was new to me) over simple cast (however harmless that seem to be).

I wouldn't rely on *harmless memory leaks* because we cannot predict the life time of each module. I think here we're only talking about C++ modules, right? I tested this short code with `cc -Wall -otest test.c` and `clang -Wall -otest test.c`:
```c
#include <stdio.h>
int main(){
    char *a = "test";
    printf("%s\n", a);
}
```
I got no warnings or errors. Then, the same code (`test.cpp`) with `c++ -Wall -otest test.cpp` and `clang++ -Wall -otest test.cpp` produced:
```
test.cpp: In function ‘int main()’:
test.cpp:3:15: warning: deprecated conversion from string constant to ‘char*’ [-Wwrite-strings]
     char *a = "test";
               ^
```
and
```
test.c:3:15: warning: conversion from string literal to 'char *' is deprecated [-Wc++11-compat-deprecated-writable-strings]
    char *a = "test";
              ^
```
But it was a matter of simple type casting to `(char *)` to get rid of these warnings.
```c
#include <stdio.h>
int main(){
    char *a = (char *)"test";
    printf("%s\n", a);
}
```
If I'm not missing something, can't we just cast it to `(char *)` explicitly only in C++ modules?

Reply:
If it would "only" be an argument to a printf I would agree totally. But there is no guarantee `Option->answer` will not be altered at some point, or is there.. ?

In any case both the string literal as well as a mutable string will be freed at r.viewshed execution finishes.

Reply:
> If it would "only" be an argument to a printf I would agree totally. But there is no guarantee `Option->answer` will not be altered at some point, or is there.. ?

I see. At this point, a lot of modules use static `char *` to assign default answers. Once parsed, they can either come directly from the default (read only) or from the user (`G_malloc` from `lib/gis/parser.c`, but is usually never explicitly freed by modules). It would be a developer's mistake if s/he treated `->answer` as modifiable when its default was not dynamically allocated.

> 
> In any case both the string literal as well as a mutable string will be freed at r.viewshed execution finishes.

Yes, any allocated memory will be freed when a module completes and it's a relatively small amount of memory.

Because `parser.c` `G_malloc` memory for user-passed options, which again is rarely explicitly freed by developers, I think `G_store` is safe and even more consistent than static `char *`.

Maybe, I just don't like asymmetry where we only allocate memory, but never free it in the *same code*. If we have to, we can always use `G_add_error_handler` to free manually allocated memory for `->answer`. Again, at the end of module execution, there is not much gain in doing so other than making us satisfied ;-).

Reply:
Implemented with `G_store()`. I believe this is the least bad solution.

Reply:
> Implemented with `G_store()`. I believe this is the least bad solution.

Looking at the code, answer is actually copied to another buffer and is never modified directly. I mean, let's consider the developer's intention. If s/he wanted to modify answer in place, `G_store` might not be a bad solution. However, if answer (default or not) never changes, simply casting will be fine. Do we want to update all C modules that do this (`->answer = "default";`)? I think not.